### PR TITLE
MGMT-6437: Redirecting cluster metrics to the correct Telemeter server.

### PR DESCRIPTION
**Pull request by ybettan**
Currently all clusters send metrics to the default Telemeter server
which is the prod instance of Telemeter in the cloud.

This is wrong for 2 reasons:

1. integration/stage clusters statistics shouldn't be a part of
   production dashboards.

2. integration/stage service is configured to work against integration/stage
   AMS therefore, for those clusters, when metrics reaches prod
   Telemeter, they will fail on authN anyway.

This change make sure each env send it metrics to the correct server or
being disabled.

Signed-off-by: Yoni Bettan <ybettan@redhat.com>

### CODE REVIEW

**Comment by ybettan** [comment]
Missing:
* subsystem-tests
* system manual test

**Comment by ybettan** [comment]
/retest
https://coreos.slack.com/archives/C014N2VLTQE/p1621509027027000?thread_ts=1621446988.016700&cid=C014N2VLTQE

File: internal/cluster/cluster.go [hunk 1]
```diff
@@ -1058,6 +1057,15 @@ func (m *Manager) GenerateAdditionalManifests(ctx context.Context, cluster *comm
 			return errors.Wrapf(err, "Cluster %s - failed to generate manifests for vmware hosts", cluster.ID.String())
 		}
 	}
+
+	if err := m.manifestsGeneratorAPI.AddTelemeterManifest(ctx, log, cluster); err != nil {
+		return errors.Wrap(err, "failed to add telemeter manifest")
+	}
+
+	if err := m.rp.operatorsAPI.GenerateManifests(ctx, cluster); err != nil {
```
Comment (eranco74) [inline internal/cluster/cluster.go:1065]: Why did you move it (it used to be before AddDisableVmwareTunnelOffloading) ?
Are you sure the order doesn't matter?

File: internal/network/manifests_generator.go [hunk 2]
```diff
@@ -340,3 +348,30 @@ func (m *ManifestsGenerator) AddDisableVmwareTunnelOffloading(ctx context.Contex
 	}
 	return nil
 }
+
+const redirectTelemeterStageManifest = `
```
Comment (eranco74) [inline internal/network/manifests_generator.go:352]: Move to the top next to all other consts

File: internal/network/manifests_generator.go [hunk 3]
```diff
@@ -24,15 +25,22 @@ type ManifestsGeneratorAPI interface {
 	AddChronyManifest(ctx context.Context, log logrus.FieldLogger, c *common.Cluster) error
 	AddDnsmasqForSingleNode(ctx context.Context, log logrus.FieldLogger, c *common.Cluster) error
 	AddDisableVmwareTunnelOffloading(ctx context.Context, log logrus.FieldLogger, c *common.Cluster) error
+	AddTelemeterManifest(ctx context.Context, log logrus.FieldLogger, c *common.Cluster) error
+}
+
+type Config struct {
+	ServiceBaseURL string `envconfig:"SERVICE_BASE_URL"`
```
Comment (eranco74) [inline internal/network/manifests_generator.go:32]: Why do you need this config?
Why not just pass the relevant value (staging/prod) to `AddTelemeterManifest`?

File: internal/network/manifests_generator.go [hunk 4]
```diff
@@ -340,3 +348,30 @@ func (m *ManifestsGenerator) AddDisableVmwareTunnelOffloading(ctx context.Contex
...
+    telemeterClient:
+	  telemeterServerURL: https://infogw.api.stage.openshift.com
+`
+
+// Default Telemeter server is prod.
+// In case the cluster is created in stage env we need to redirct to Telemter-stage
+// Note: There is no Telemeter-integraion so we won't redirect and we will let it fail on authN when reaching the
+// 		 wrong Telemeter server. We will do the same for all non-cloud envs
+func (m *ManifestsGenerator) AddTelemeterManifest(ctx context.Context, log logrus.FieldLogger, c *common.Cluster) error {
+	if strings.Contains(m.Config.ServiceBaseURL, "stage") {
```
Comment (eranco74) [inline internal/network/manifests_generator.go:369]: I think that instead of spamming the prod Telemeter auth service it will be better to use a dummy URL instead.

File: internal/cluster/cluster.go (hunk 1 above)
Comment (ybettan) [reply internal/cluster/cluster.go:1065]: No special reason, just so they are grouped by API (operators manifests comes from a different API than network manifests).

This is just file generation no?! I don't think the order matter.

File: internal/network/manifests_generator.go (hunk 3 above)
Comment (ybettan) [reply internal/network/manifests_generator.go:32]: Not sure I get what you mean, the caller doesn't now this value, it should be passed through the whole stack from bminventory pkg.

See https://coreos.slack.com/archives/CUPJTHQ5P/p1621515876386800

File: internal/network/manifests_generator.go (hunk 2 above)
Comment (ybettan) [reply internal/network/manifests_generator.go:352]: They aren't all grouped on top of the file, for example in `AddDisableVmwareTunnelOffloading` the coasts are defined just before the method.

I don't mind moving it to the top though.

File: internal/network/manifests_generator.go (hunk 4 above)
Comment (ybettan) [reply internal/network/manifests_generator.go:369]: You mean in all other cases which are not stage/prod to create a ConfigMap that redirect the metrics to a dummy URL? (instead of just letting them fail on auth when reaching prod ?)

Comment (ybettan) [reply internal/network/manifests_generator.go:369]: I am not sure how the cluster will react to a dummy URL. Since I don't have a real way to test it until it reaches stage I rather keep the default behaviour for all envs that we can't control.

File: internal/cluster/cluster.go (hunk 1 above)
Comment (eranco74) [reply internal/cluster/cluster.go:1065]: I don't know what it does, so now we need to check...
In essence, if you don't have a reason to change something - don't.

File: internal/network/manifests_generator.go (hunk 4 above)
Comment (eranco74) [reply internal/network/manifests_generator.go:369]: You can install a cluster and create the above configmap with a dummy URL.

File: internal/network/manifests_generator_test.go [hunk 5]
```diff
@@ -305,3 +305,52 @@ var _ = Describe("dnsmasq manifest", func() {
...
+		}
+		Expect(db.Create(&cluster).Error).NotTo(HaveOccurred())
+	})
+
+	AfterEach(func() {
+		ctrl.Finish()
+		common.DeleteTestDB(db, dbName)
+	})
+
+	It("happy flow", func() {
```
Comment (eranco74) [inline internal/network/manifests_generator_test.go:345]: Please add a test case with  ServiceBaseURL that doesn't have the word `stage`

File: internal/network/manifests_generator.go (hunk 4 above)
Comment (eranco74) [reply internal/network/manifests_generator.go:369]: Also, you can install from assisted-test-infra it should act just the same (add dummy URL)
The only thing you can't test until you get to staging is the logic you currently have.

File: internal/cluster/cluster.go (hunk 1 above)
Comment (ybettan) [reply internal/cluster/cluster.go:1065]: I will fix the order to fit the original order then. Thanks for the tip.

File: internal/network/manifests_generator.go (hunk 4 above)
Comment (ybettan) [reply internal/network/manifests_generator.go:369]: I will do it.

Comment (romfreiman) [reply internal/network/manifests_generator.go:369]: `// Note: There is no Telemeter-integraion so we won't redirect and we will let it fail on authN when reaching the// 		 wrong Telemeter server. We will do the same for all non-cloud envs`
Please DONT! What if we run a scale test, will it DDOS prod telemetry?

Comment (romfreiman) [reply internal/network/manifests_generator.go:369]: btw, it can be completely disabled:
https://docs.openshift.com/container-platform/4.7/support/remote_health_monitoring/opting-out-of-remote-health-reporting.html#insights-operator-new-pull-secret_opting-out-remote-health-reporting

**Review by romfreiman — CHANGES_REQUESTED** [review]

File: internal/network/manifests_generator.go (hunk 4 above)
Comment (ybettan) [reply internal/network/manifests_generator.go:369]: @romfreiman This was the default behaviour until today, nothing has changed - all scale tests until today had this exact behaviour, I just described it in the comment - this is not new code.
I am aware of https://docs.openshift.com/container-platform/4.7/support/remote_health_monitoring/opting-out-of-remote-health-reporting.html#insights-operator-new-pull-secret_opting-out-remote-health-reporting but this will require modifying manifests created by openshift-installer (not just adding new custom manifests) - I don't think this is the solution we seek.

Comment (ybettan) [reply internal/network/manifests_generator.go:369]: Anyway it is getting changed to @eranco74 suggestion (using dummy URL)

File: internal/network/manifests_generator_test.go (hunk 5 above)
Comment (ybettan) [reply internal/network/manifests_generator_test.go]: Done.

Comment (eranco74) [reply internal/network/manifests_generator_test.go]: Where is the test?

**Review by eranco74 — COMMENTED** [review]
/lgtm

**Comment by ybettan** [comment]
We can see in the service that the custom manifest was created:

```
$ curl -s $(minikube service assisted-service -n assisted-installer --url)/api/assisted-install/v1/clusters/c12773ed-5bcf-407c-8b4b-bf23808b9c86/manifests | jq '.'
[
  {
    "file_name": "masters-chrony-configuration.yaml",
    "folder": "openshift"
  },
  {
    "file_name": "redirect-telemeter.yaml",
    "folder": "openshift"
  },
  {
    "file_name": "workers-chrony-configuration.yaml",
    "folder": "openshift"
  }
]
```

In the created cluster we can see that the ConfigMap was applied with the correct URL:
```
$ oc --kubeconfig kubeconfig describe cm/cluster-monitoring-config -n openshift-monitoring
Name:         cluster-monitoring-config
Namespace:    openshift-monitoring
Labels:       <none>
Annotations:  <none>

Data
====
config.yaml:
----
telemeterClient:
  telemeterServerURL: https://dummy.com

Events:  <none>
```

**Review by gamli75 — APPROVED** [review]
